segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/CacheStorage.java

```diff
@@ -0,0 +1,122 @@
+     * @throws CacheFullException If there is no more capacity in the {@link CacheStorage} to accommodate this entry.
+     * @throws CacheException     If another unexpected exception occurred.
+     */
+    int replace(int address, @NonNull BufferView data);
+
+    /**
+     * Gets the number of bytes that can be appended to an entry that has the given length.
+     *
+     * @param currentLength The current length of the entry.
+     * @return An non-negative integer representing the number of bytes that can be appended.
+     */
+    int getAppendableLength(int currentLength);
+
+    /**
+     * Performs an atomic compare-and-appends of data at the end of an entry. This operation will only take effect if
+     * the current length of the entry matches the argument `expectedLength` and both the entry's length and contents
+     * will be updated atomically.
+     *
+     * @param address        An integer representing the address to append to.
+     * @param expectedLength The currently known length of this entry.
+     * @param data           A {@link BufferView} representing the data to append.
+     * @return An integer representing the number of bytes that were appended. If 0, then the entry at the given address
+     * is already at its max capacity (its length equals {@link #getMaxEntryLength()}.
+     * @throws IncorrectCacheEntryLengthException If `expectedLength` does not match the current length of the entry.
+     * @throws IllegalArgumentException If address does not point to a valid entry or the given {@link BufferView} is
+     * too long to fit into the given entry (use {@link #getAppendableLength} to determine maximum length).
+     */
+    int append(int address, int expectedLength, @NonNull BufferView data);
+
+    /**
+     * Deletes the entry at the given address. If this address does not currently map to an entry, this operation will have
+     * no effect.
+     *
+     * @param address An integer representing the address to delete.
+     */
+    void delete(int address);
+
+    /**
+     * Retrieves the contents of an entry with the given address.
+     *
+     * @param address An integer representing the address to retrieve.
+     * @return A read-only {@link BufferView} that can be used to access the data, or null if no entry is mapped to this
+     * address.
+     */
+    BufferView get(int address);
+
+    /**
+     * Returns a {@link CacheState} representing the current state of the {@link CacheStorage}.
+     *
+     * @return A new {@link CacheState} instance.
+     */
+    CacheState getState();
+
+    /**
+     * Sets a callback that will be invoked during {@link #insert} if there is insufficient capacity to add more entries.
+     *
+     * @param cacheFullCallback The callback to invoke. This should return `true` if a cache cleanup was performed, and
+     *                          `false` otherwise.
+     */
+    void setCacheFullCallback(@NonNull Supplier<Boolean> cacheFullCallback);
```

Comment:
The eviction policy is implemented by `CacheManager` in conjunction with all the other information that the `ReadIndex` and other cache clients have. This is purely a Cache Storage: it stores and provides access to data, but does not bother with eviction as it has insufficient information to decide what can be evicted.

In other cases, we could write a small module that can manage this Cache Storage and perform eviction based on policies.

The point of this method is to notify the `CacheManager` that the cache is full and that it should try to execute an eviction if possible, otherwise it will reject new inserts with `CacheFullException`.

Reply:
I think you are confirming that a policy is triggered by such a callback. When the cache is full, it needs to evict due to reaching its capacity, and some policy needs to make a call on what to evict. You are saying that the storage part  notifies via this callback and whatever policy is implemented, if any at all, reacts to such a notification. In the current implementation, this is the cache manager.

Reply:
Yes that is correct. The Cache Storage has insufficient information about what is stored and what should be evicted. The Cache Manager (external) has all the info it needs (what data is in Tier 2, for example) to make such decisions. The Cache Manager is also wired into the Throttler to slow ingestion down if it cannot evict fast enough.